### PR TITLE
api: Enhance copyObject to support metadata.

### PR DESCRIPTION
**Pull request by harshavardhana**
This change allows copyObject() to replace metadata
with new metadata on the destination object. Additionally
this also allows for replacing metadata on existing objects
as well.

### CODE REVIEW

File: api/src/main/java/io/minio/CopyConditions.java [hunk 1]
```diff
@@ -89,6 +89,20 @@ public void setMatchETagNone(String etag) throws InvalidArgumentException {
...
   /**
+   * Set metadata directive to specify whether the metadata
+   * is copied from the source object or replaced with metadata
+   * provided in the request.
+   */
+  public void setMetadataDirective(String value) {
+    if (value == null) {
+      // http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html
+      // Default value is - COPY.
+      value = "COPY";
```
Comment (balamurugana) [inline api/src/main/java/io/minio/CopyConditions.java]: Can you add respective enum typeto denote the value? I see only `COPY` and `REPLACE` are permitted.

File: functional/FunctionalTest.java [hunk 2]
```diff
@@ -1391,6 +1391,44 @@ public static void copyObject_test7() throws Exception {
...
+                        copyConditions, metadata);
+    } catch (ErrorResponseException e) {
+      // File should not be copied as object was modified after date set in copyConditions.
+      if (!e.errorResponse().code().equals("PreconditionFailed")) {
+        throw e;
+      }
+    }
+
+    ObjectStat objectStat = client.statObject(bucketName, filename);
+    if (!customContentType.equals(objectStat.contentType())) {
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Can you add test for `COPY` metadata directive?

File: api/src/main/java/io/minio/MinioClient.java [hunk 3]
```diff
@@ -3158,58 +3209,58 @@ public void putObject(String bucketName, String objectName, InputStream stream,
...
    *
    * @see #putObject(String bucketName, String objectName, String fileName, String contentType)
-   * 
+   *
    * @see #putObject(String bucketName, String objectName, String fileName)
-   * 
+   *
    * @see #putObject(String bucketName, String objectName, InputStream stream, long size, String contentType)
-   * 
+   *
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: I think this automatic in my `emacs` not sure if i can turn this off..

File: functional/FunctionalTest.java [hunk 4]
```diff
@@ -692,7 +692,7 @@ public static void listObject_test5() throws Exception {
 
     // Check the number of uploaded objects
     if (i != objCount) {
-      throw new Exception("[FAILED] Test: recursive: listObject_test5(), number of items, expected: " 
+      throw new Exception("[FAILED] Test: recursive: listObject_test5(), number of items, expected: "
```
Comment (harshavardhana) [inline functional/FunctionalTest.java]: As explained above..

File: functional/FunctionalTest.java (hunk 2 above)
Comment (harshavardhana) [inline functional/FunctionalTest.java]: COPY is default @balamurugana so didn't think that was necessary.

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -1877,7 +1879,8 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
         NoResponseException, ErrorResponseException, InternalException, IOException, XmlPullParserException,
         InvalidArgumentException {
 
-    copyObject(bucketName, objectName, destBucketName, null, copyConditions);
+    copyObject(bucketName, objectName, destBucketName, null, copyConditions,
+               null);
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: This is not a cleanup @balamurugana it is reflecting the change in function signature.

File: api/src/main/java/io/minio/MinioClient.java [hunk 6]
```diff
@@ -1945,6 +1993,9 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
     if (copyConditions != null) {
       headerMap.putAll(copyConditions.getConditions());
     }
+    if (metadata != null) {
+      headerMap.putAll(Collections.unmodifiableMap(metadata));
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: Kept it similar to getConditions()

File: api/src/main/java/io/minio/CopyConditions.java (hunk 1 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/CopyConditions.java]: The reason i didn't add enum is because it wasn't needed,  server returns an error anyways if the wrong metadata directive was set.  Doing an enum disadvantage is that we are defining new types where none exists from a server perspective. 

Users of aws-sdk would have a hard time looking up these enums. Perhaps a strings constant is better..

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Got it.   The line could fit into 120 char long support.   The breakage makes me think of a cleanup.

File: api/src/main/java/io/minio/MinioClient.java (hunk 6 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Yep. No need to wrap.

File: api/src/main/java/io/minio/MinioClient.java (hunk 3 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: My `emacs` doesn't do automatically.  I think there could be a way to do so.

File: api/src/main/java/io/minio/CopyConditions.java (hunk 1 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/CopyConditions.java]: Adding a new enum anyways. to keep it similar in behavior to other types of similar code.

Comment (balamurugana) [inline api/src/main/java/io/minio/CopyConditions.java]: As we provide a method to set a value for `x-amz-metadata-directive` header, its required to validate given arg.  String constants are not java style.  enum provides a strict typing.  By looking into javadoc, everybody knows what to pass to the method.

File: api/src/main/java/io/minio/MinioClient.java (hunk 3 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: Will see..

Comment (harshavardhana) [inline api/src/main/java/io/minio/MinioClient.java]: Its basically whitespace cleanup.

Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Unwanted line modification makes review harder.  Its always recommended to send a separate PR for such cleanup than mingling with new API addition.

File: api/src/main/java/io/minio/CopyConditions.java [hunk 7]
```diff
@@ -24,13 +24,13 @@
 import org.joda.time.DateTime;
 
 import io.minio.errors.InvalidArgumentException;
+import io.minio.MetadataDirective;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/CopyConditions.java]: Its necessary in same package.

File: api/src/main/java/io/minio/CopyConditions.java [hunk 8]
```diff
@@ -89,6 +89,20 @@ public void setMatchETagNone(String etag) throws InvalidArgumentException {
   }
 
   /**
+   * Set metadata directive to specify whether the metadata
+   * is copied from the source object or replaced with metadata
+   * provided in the request.
+   */
+  public void setMetadataDirective(MetadataDirective value) {
+    if (value == null) {
```
Comment (balamurugana) [inline api/src/main/java/io/minio/CopyConditions.java]: I don't think its necessary for an enum argument.

File: api/src/main/java/io/minio/MetadataDirective.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+  private MetadataDirective(String value) {
+    this.value = value;
+  }
+
+  /**
+   * Returns MetadataDirective enum of given string.
+   */
+  public static MetadataDirective fromString(String md) {
+    if (md == null) {
+      throw new IllegalArgumentException("null md");
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MetadataDirective.java]: the error supposes to be `null metadata value`

File: api/src/main/java/io/minio/CopyConditions.java (hunk 8 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/CopyConditions.java]: Two kinds of implementation:
1. the current style.  user is required to pass correct enum even though copy is default.
2. another style is `setReplaceMetadataDirective()`.  This will add the header with `REPLACE` value. No enum necessary.

My choice is less code.  But its your call.

File: api/src/main/java/io/minio/CopyConditions.java (hunk 7 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/CopyConditions.java]: you want me to move this to CopyCondtions?

File: api/src/main/java/io/minio/CopyConditions.java (hunk 8 above)
Comment (harshavardhana) [inline api/src/main/java/io/minio/CopyConditions.java]: Like the setReplaceMetadataDirective behavior.

**Comment by harshavardhana** [comment]
@balamurugana can you approve this? if nothing is pending.
